modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -846,52 +846,127 @@ def join(
+                            if col == df._index_name(c):
+                                new_name = c
+                                break
+                    if new_name is None:
+                        raise ValueError(f"'{what}' references unknown column {col}")
+                    if new_col_names is col_names:
+                        new_col_names = col_names.copy()
+                    new_col_names[i] = new_name
+            return df, new_col_names
+
+        orig_left_on = left_on
+        orig_right_on = right_on
+        left, left_on = validate("left_on", self, left_on)
+        right, right_on = validate("right_on", other, right_on)
         new_dtypes = []
         exprs = OrderedDict()
 
-        left_conflicts = set(self.columns) & (set(other.columns) - set(right_on))
-        right_conflicts = set(other.columns) & (set(self.columns) - set(left_on))
-        conflicting_cols = left_conflicts | right_conflicts
-        for c in self.columns:
-            new_name = f"{c}{suffixes[0]}" if c in conflicting_cols else c
-            new_columns.append(new_name)
-            new_dtypes.append(self._dtypes[c])
-            exprs[new_name] = self.ref(c)
-        for c in other.columns:
-            if c not in left_on or c not in right_on:
-                new_name = f"{c}{suffixes[1]}" if c in conflicting_cols else c
-                new_columns.append(new_name)
-                new_dtypes.append(other._dtypes[c])
-                exprs[new_name] = other.ref(c)
+        # Join by index
+        if (left_on is not orig_left_on) or (right_on is not orig_right_on):
+
+            def to_empty_pandas_df(df):
+                idx = df._index_cache
+                if idx is not None:
+                    idx = idx[0:1]
+                elif df._index_cols is not None:
+                    if len(df._index_cols) > 1:
+                        arrays = [[i] for i in range(len(df._index_cols))]
+                        names = [df._index_name(n) for n in df._index_cols]
+                        idx = pd.MultiIndex.from_arrays(arrays, names=names)
+                return pd.DataFrame(columns=df.columns, index=idx)
+
+            merged = to_empty_pandas_df(self).merge(
+                to_empty_pandas_df(other),
+                how=how,
+                left_on=orig_left_on,
+                right_on=orig_right_on,
+                sort=sort,
+                suffixes=suffixes,
+            )
 
-        condition = self._build_equi_join_condition(other, left_on, right_on)
+            if len(merged.index.names) == 1 and (merged.index.names[0] is None):
+                index_cols = None
+            else:
+                index_cols = left._mangle_index_names(merged.index.names)
+                for n, m in zip(merged.index.names, index_cols):
+                    df = left if m in left._dtypes else right
+                    exprs[n] = df.ref(m)
+                    new_dtypes.append(df._dtypes[m])
```

Comment:
could you please elaborate on this? why are we checking dtypes here?

Reply:
Most probably, we can find dtype by the index col name in the left frame, but I'm not sure how the Pandas index merge works, so added this expression for the right frame.